src/lib/bootloader/grub_install.rb

```diff
@@ -2,30 +2,72 @@
 require "yast2/execute"
 
 Yast.import "Arch"
+Yast.import "Report"
 
 module Bootloader
   # Wraps grub install script for easier usage.
   class GrubInstall
+    include Yast::Logger
+    include Yast::I18n
+
     def initialize(efi: false)
       @efi = efi
+      textdomain "bootloader"
     end
 
+    # Runs grub2 install command.
+    #
+    # @param devices[Array<String>] list to which grub2 should be installed.
+    #   Ignored when grub2 does not need device.
+    # @param secure_boot [Boolean] if secure boot variant should be used
+    # @param trusted_boot [Boolean] if trusted boot variant should be used
+    # @return [Array<String>] list for which install failed
     def execute(devices: [], secure_boot: false, trusted_boot: false)
       raise "cannot have secure boot without efi" if secure_boot && !efi
 
       cmd = basic_cmd(secure_boot, trusted_boot)
 
       if no_device_install?
         Yast::Execute.on_target(cmd)
+        []
       else
-        devices.each { |d| Yast::Execute.on_target(cmd + [d]) }
+        return [] if devices.empty?
+        last_failure = nil
+        res = devices.select do |device|
+          begin
+            Yast::Execute.on_target!(cmd + [device])
+            false
+          rescue Cheetah::ExecutionFailed => e
+            log.warn "Failed to install grub to device #{device}. #{e.inspect}"
+            last_failure = e
+            true
+          end
+        end
+
+        # Failed to install to all devices
+        report_failure(last_failure) if res.size == devices.size
```

Comment:
so it can fail for all devices but reports only one? It is a bit strange in light of fact that error is reported *only* when grub could not be written to all devices.

Reply:
in logs you see all failures, but reported is only last one as usually it is just single reason for all devices ( like missing bios grub, missing grub2 at all or some other issues ). So goal is to report user that there is a error ( so we do not hide that we have problem to make system bootable ), but it is not shown for all failures.

Reply:
well, I meant that just a report "Failed to install grub." would be enough instead of appending a device name ... but that's only NP for me.